src/Hamiltonians/HOCartesianCentralImpurity.jl

```diff
@@ -0,0 +1,200 @@
+harmonic oscillator function evaluated at the origin. The overall sign is
+determined when the matrix element is evaluated in [`ho_delta_potential`](@ref).
+"""
+function log_abs_oscillator_zero(n)
+    isodd(n) && return -Inf, 1.0
+    x, _ = SpecialFunctions.logabsgamma((1-n)/2)
+    y, _ = SpecialFunctions.logabsgamma(n+1)
+    result = log(pi)/4 + (n/2)*log(2) - x - y/2
+    return result
+end
+
+"""
+    ho_delta_potential(S, i, j; vals)
+
+Returns the product of two harmonic oscillator functions 
+evaluated at the origin, 
+```math
+    v_{ij} = \\phi_\\mathbf{n}_i(0) \\phi_\\mathbf{n}_j(0)
+```
+which is only non-zero for even-parity states.
+Mode indices `i,j` map to ``D``-dimensional tuples of harmonic oscillator 
+indices ``\\mathbf{n}_i``. `S` defines the bounds of the Cartesian HO modes.
+The values ``\\phi_i(0)`` are calculated with [`log_abs_oscillator_zero`](@ref) 
+and should be precomputed by [`HOCartesianCentralImpurity`](@ref) and passed 
+in as the vector `vals`.
+"""
+function ho_delta_potential(S, i, j; 
+    vals = [log_abs_oscillator_zero(k) for k in 0:2:(maximum(S)-1)]
+    )
+    states = CartesianIndices(S)
+    ho_indices = (Tuple(states[i])..., Tuple(states[j])...) .- 1
+    if all(iseven.(ho_indices))
+        doubly_evens = count(iseven.(ho_indices .÷ 2))
+        sign = (-1)^doubly_evens
+        return sign * exp(sum(vals[k÷2 + 1] for k in ho_indices))
+    else
+        return 0.0
+    end
+end
+
+"""
+    HOCartesianCentralImpurity(addr; kwargs...)
+
+Hamiltonian of non-interacting particles in an arbitrary harmonic trap with a delta-function 
+potential at the centre, with strength `g`,
+```math
+\\hat{H}_\\mathrm{rel} = \\sum_\\mathbf{i} ϵ_\\mathbf{i} n_\\mathbf{i} 
+    + g\\sum_\\mathbf{ij} V_\\mathbf{ij} a^†_\\mathbf{i} a_\\mathbf{j}.
+```
+For a ``D``-dimensional harmonic oscillator indices ``\\mathbf{i}, \\mathbf{j}, \\ldots`` 
+are ``D``-tuples. The energy scale is defined by the first dimension i.e. ``\\hbar \\omega_x`` 
+so that single particle energies are 
+```math
+    \\frac{\\epsilon_\\mathbf{i}}{\\hbar \\omega_x} = (i_x + 1/2) + \\eta_y (i_y+1/2) + \\ldots.
+```
+The factors ``\\eta_y, \\ldots`` allow for anisotropic trapping geometries and are assumed to 
+be greater than `1` so that ``\\omega_x`` is the smallest trapping frequency.
+
+Matrix elements ``V_{\\mathbf{ij}}`` are for a delta function potential calculated in this basis
+```math
+    V_{\\mathbf{ij}} = \\prod_{d \\in x, y,\\ldots} \\psi_{i_d}(0) \\psi_{j_d}(0).
+```
+Only even parity states feel this impurity, so all ``i_d`` are even. Note that the matrix 
+representation of this Hamiltonian is completely dense in the even-parity subspace.
+
+# Arguments
+
+* `addr`: the starting address, defines number of particles and total number of modes.
+* `M=num_modes(addr) - 1`: the maximum harmonic oscillator index number in the ``x``-dimension.
+    Must be even. Index number for the harmonic oscillator groundstate is `0`.
```

Comment:
Hmm, could we pass the number of modes here rather than the largest index when starting the counting from 0? This way you would not have to subtract 1 and this is also more consistent with working with the number of modes elsewhere.

Furthermore, can we please rename this keyword argument to `x_modes` or some other more descriptive name? I find `M` confusing as elsewhere it is used for the total number of modes.

Reply:
I've renamed it to `max_nx`. 

You would still have to subtract 1 somewhere. It is natural to index HO states with 0 and it makes handling the aspect ratio nicer. Also, it would then have to enforce that `max_nx` is odd, which would be confusing after saying that only even-parity states are involved.